### PR TITLE
adds start view and end view to resetEpoch

### LINKED ISSUE #180: Accept FirstView, FinalView as arguments to resetEpoch
## Context

When resetting an epoch, it is critical that the smart contract epoch metadata remains in sync (exactly) with the protocol state epoch metadata. Since the protocol state epoch metadata is generated first, then injected to the smart contract with the `resetEpoch` function, the `resetEpoch` function must accept as input all metadata fields which must be synchronized. 

## Definition of Done

* Accept 2 additional inputs to `resetEpoch` - `startView` and `endView`
* Change computation of `stakingEndView` to use injected `startView` rather than current block 
Change  to use startView that is injected
* Enforce sanity check for epoch length input, given the configured DKG phase length and staking auction length, similar to the [checks in place for changing configuration](https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowEpoch.cdc#L275)

### PR DESCRIPTION
**Pull request by joshuahannan**
Closes #180 

Adds `startView` and `endView` as parameters to the `resetEpoch` method

Open question: Should there be any other pre-conditions or expectations for these parameters?

### CODE REVIEW

**Review by jordanschalm — COMMENTED** [review]
Hopefully [this](https://github.com/onflow/flow-core-contracts/issues/180#issuecomment-847951497) answers the Q about the sanity check. Otherwise looks great!

File: contracts/epochs/FlowEpoch.cdc [hunk 1]
```diff
@@ -377,9 +377,12 @@ pub contract FlowEpoch {
         pub fun resetEpoch(
             randomSource: String,
             newPayout: UFix64?,
+            startView: UInt64,
+            endView: UInt64,
             collectorClusters: [FlowEpochClusterQC.Cluster],
             clusterQCs: [FlowEpochClusterQC.ClusterQC]
-            dkgPubKeys: [String]) {
+            dkgPubKeys: [String])
+        {
```
Comment (jordanschalm) [inline contracts/epochs/FlowEpoch.cdc:386]:
```suggestion
        {
        pre {
                FlowEpoch.isValidPhaseConfiguration(FlowEpoch.configurableMetadata.numViewsInStakingAuction, FlowEpoch.configurableMetadata.numViewsInDKGPhase, endView-startView+1)
        }
```

**Comment by joshuahannan** [comment]
@jordanschalm I added the pre-condition and a test in 46734ba

File: lib/go/test/flow_epoch_test.go [hunk 2]
```diff
@@ -854,11 +854,38 @@ func TestEpochReset(t *testing.T) {
 
 	})
 
+	t.Run("Cannot reset the epoch if the phase endView isn't high enough", func(t *testing.T) {
```
Comment (jordanschalm) [inline lib/go/test/flow_epoch_test.go:857]: 🙇

**Review by jordanschalm — APPROVED** [review]

**Pull request merged**